bin/init_exercise.py

```diff
@@ -0,0 +1,583 @@
+    sys.exit(1)
+
+# check version info
+if sys.version_info[0] != 3 or sys.version_info[1] < 5:
+    print("This script requires Python 3.5 or higher", file=sys.stderr)
+    # exiting like this isn't great for library use, but at least it's a quick fail
+    sys.exit(1)
+
+
+def output_of(cmd, check_returncode=True):
+    "Return the stdout of the given command"
+    sp = subprocess.run(shlex.split(cmd),
+                        stdout=subprocess.PIPE,
+                        universal_newlines=True)
+    if check_returncode:
+        sp.check_returncode()
+    return sp.stdout.strip()
+
+
+REPO_ROOT = output_of('git rev-parse --show-toplevel')
+EXERCISES = os.path.join(REPO_ROOT, 'exercises')
+ITEM_NAME_CHARS = {c for c in string.ascii_lowercase + string.digits + '_'}
+VERSION = "0.1.0"
+
+
+def to_item_name(description):
+    "Produce a valid rust item name from arbitrary inputs"
+    item = description.lower().replace(' ', '_')
+    item = [c for c in item if c in ITEM_NAME_CHARS]
+    while len(item) > 0 and item[0] in string.digits:
+        item = item[1:]
+    if len(item) == 0:
+        raise ValueError("Could not produce an item name from " + description)
+    return ''.join(item)
+
+
+def to_crate_name(name):
+    return name.replace('-', '_')
+
+
+def url_for(name, file):
+    return (
+        "https://raw.githubusercontent.com/exercism/problem-specifications"
+        f"/master/exercises/{name}/{file}"
+    )
+
+
+def get_problem_specification(name):
+    """
+    Try to get problem specifications for the exercise of the given name.
+
+    If the problem specifications repo doesn't exist or the exercise does not
+    exist within the specifications repo, returns None.
+    Otherwise, returns a dict, of which the values might be None or str.
+    """
+    try:
+        with urllib.request.urlopen(url_for(name, 'canonical-data.json')) as response:
+            return json.loads(response.read())
+    except (urllib.request.URLError, json.JSONDecodeError):
+        pass
+
+
+@contextmanager
+def inside(path):
+    cwd = os.getcwd()
+    os.chdir(path)
+    try:
+        yield
+    finally:
+        os.chdir(cwd)
+
+
+def make_exercise(name, use_maplit):
+    "Make a new exercise with the specified name"
+    with inside(EXERCISES):
+        if os.path.exists(name):
+            print(f"{name} already exists; aborting", file=sys.stderr)
+            sys.exit(1)
+        subprocess.run(['cargo', 'new', name])
+    exercise_dir = os.path.join(EXERCISES, name)
+    # blank out the default lib.rs
+    with inside(exercise_dir):
+        with open('.gitignore', 'w') as gitignore:
+            print("Cargo.lock  # We're building a library, not a binary", file=gitignore)
+            print("            # http://doc.crates.io/faq.html#why-do-binaries-have-cargolock-in-version-control-but-not-libraries", file=gitignore)
+        with open(os.path.join('src', 'lib.rs'), 'w') as lib_rs:
+            lib_rs.truncate()

```

Comment:
excuse me that this is the only comment I give right now instead of a full review, as my allotted time is up:

I meant https://docs.python.org/3/library/os.html#os.truncate such that the file no longer needs to be opened

But I'm not sure that actually makes a difference, so I actually don't care.

Reply:
Huh, ok. It still works, so I suppose it's just a case where there are a few valid paths to accomplish a goal in the standard library.